rfc/0000-batch-job-groups.rst

```diff
@@ -0,0 +1,1179 @@
+adding new features in the future or planning a future rewrite of the
+entire system. The UI will also need to be more complicated when we
+expose a nested directory hierarchy to the users.
+
+The proposed simplifications to how cancellation are done will
+increase future developer productivity as this has always been a
+tricky and confusing part of our system especially with how it relates
+to always_run jobs.
+
+
+------------
+Alternatives
+------------
+
+The existing workaround QoB uses when waiting on a wave of worker jobs
+to complete is to poll for when the number of completed jobs is equal
+to the number of jobs in the batch minus 1 to compensate for the
+driver job. This logic is not straightforward. There are no existing
+workarounds for a driver job to be able to cancel a wave of worker
+jobs without cancelling itself.
+
+An alternate design to my proposed change has already been piloted and
+influenced the current design. The alternate design is a batch is the
+root job group in a job group tree and all operations on batches are
+implemented in terms of job groups. Jobs can only belong to one job
+group that is a node in the job group tree. The user assigns jobs to a
+job group in a path-like structure. All tables that were parameterized
+by batch ID are now parameterized by batch ID and job group ID. In the
+long run, this design is not as flexible as allowing users to assign
+jobs to multiple job groups or select jobs into a job group using an
+arbitrary query. This plan is more costly to implement due to making
+sure all of the the database transformations are correct. There are
+also more complicated SQL queries with using the job tree data
+structure to be able to correctly propagate billing and job state
+information up the tree and cancellation down the tree. The benefits
+of this approach are there are less edge cases and code paths to worry
+about with regards to job group creation and there is simplicity in a
+job group being analogous in implementation to how batches are
+implemented in the current system and that a job can only belong to
+one job group. Ultimately, I decided the proposed approach will be
+easier and quicker to get implemented and merged into the code base
+and will be more flexible for future use cases despite it being a
+bigger change to how our current system works than the explicit job
+group tree proposal -- consistent with feedback I got on the original
+proposal.
+
+We could also implement job groups where a job is assigned to a single
+arbitrary job group with no notion of hierarchy. The implementation
+would be very similar to what I have proposed although the assumption
+that a job belongs to at most one job group does make the SQL queries
+simpler. I can see this as an intermediate step to get to the full
+vision, but I want to make sure that if we commit to this approach
+that it does not impede the longer term vision I have outlined above.
+
+
+--------------------
+Unresolved Questions
```

Comment:
In general, an RFC should resolve the questions it answers. This is fine in a draft, but the review process needs to answer to dismiss these questions.